crates/storage/provider/src/providers/snapshot/jar.rs

```diff
@@ -0,0 +1,222 @@
+use super::LoadedJarRef;
+use crate::{BlockHashReader, BlockNumReader, HeaderProvider, TransactionsProvider};
+use reth_db::{
+    table::{Decompress, Table},
+    HeaderTD,
+};
+use reth_interfaces::{provider::ProviderError, RethResult};
+use reth_nippy_jar::NippyJarCursor;
+use reth_primitives::{
+    snapshot::SegmentHeader, Address, BlockHash, BlockHashOrNumber, BlockNumber, ChainInfo, Header,
+    SealedHeader, TransactionMeta, TransactionSigned, TransactionSignedNoHash, TxHash, TxNumber,
+    B256, U256,
+};
+use std::ops::{Deref, RangeBounds};
+
+/// Provider over a specific `NippyJar` and range.
+#[derive(Debug)]
+pub struct SnapshotJarProvider<'a>(LoadedJarRef<'a>);
+
+impl<'a> Deref for SnapshotJarProvider<'a> {
+    type Target = LoadedJarRef<'a>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<'a> From<LoadedJarRef<'a>> for SnapshotJarProvider<'a> {
+    fn from(value: LoadedJarRef<'a>) -> Self {
+        SnapshotJarProvider(value)
+    }
+}
+
+impl<'a> SnapshotJarProvider<'a> {
+    /// Provides a cursor for more granular data access.
+    pub fn cursor<'b>(&'b self) -> RethResult<NippyJarCursor<'a, SegmentHeader>>
+    where
+        'b: 'a,
+    {
+        Ok(NippyJarCursor::new(self.value(), Some(self.mmap_handle()))?)
+    }
+}
+
+impl<'a> HeaderProvider for SnapshotJarProvider<'a> {
+    fn header(&self, block_hash: &BlockHash) -> RethResult<Option<Header>> {
+        // WIP
+        let mut cursor = self.cursor()?;
+
+        let header = Header::decompress(
+            cursor.row_by_key_with_cols::<0b01, 2>(&block_hash.0).unwrap().unwrap()[0],
+        )
+        .unwrap();
```

Comment:
these should be errors, but fine with cleaning this up separately

Reply:
yeah, next PR is actually implementing the HeaderProvider, and will clean this up